docs/configuring_subprogram_specification_statement_rules.rst

```diff
@@ -0,0 +1,150 @@
+            return type_mark
+
+    designator ::= identifier | operator_symbol
+
+    formal_parameter_list ::= parameter_interface_list
+
+    interface_list ::=
+        interface_element { ; interface_element }
+
+    interface_element ::= interface_declaration
+
+There are separate rules for the structure of the subprogram specification and the indenting.
+Both rules are required to ensure proper formatting of subprogram specification statements.
+
+There are several options to the structure rules:
+
+.. |values| replace::
+   :code:`add_new_line`, :code:`remove_new_line`, :code:`ignore`
+
+.. |values2| replace::
+   :code:`remove_new_line`, :code:`ignore`
+
+.. |values3| replace::
+   :code:`yes`, :code:`no`
+
+.. |no| replace::
+   :code:`no`
+
+.. |green_diamond| image:: img/green_diamond.png
+
+.. |red_penta_star| image:: img/red_penta_star.png
+
+.. |orange_triangle| image:: img/orange_triangle.png
+
+.. |add_new_line| replace::
+   The setting :code:`add_new_line` enforces a carriage return (alias "new line") [and, consequently by indentation rules kicking in, also (indirectly) enforces the indentation of the new line]
+
+.. |remove_new_line| replace::
+   The setting :code:`remove_new_line` enforces the removal of any potential space and carriage return
+
+.. |ignore| replace::
+   The setting :code:`ignore` disables the option and hence no formatting check is done at all: spaces and new lines can be anything
+
+.. |default_add_new_line| replace::
+   :code:`add_new_line`
+
+.. |default_remove_new_line| replace::
+   :code:`remove_new_line`
+
+.. |default_ignore| replace::
+   :code:`ignore`
+
+.. |ignore_single_line| replace::
+   :code:`ignore_single_line`
+
+.. |ignore_single_line__yes| replace::
+   :code:`yes` = Ignore single line expressions.
+
+.. |ignore_single_line__no| replace::
+   :code:`no` =  Apply rules to single line expressions.
+
++----------------------------------+-----------+---------------------+---------------------------+-----------------------------+
+| Option                           | Values    | Structural Element  | Default Value             | Description                 |
++==================================+===========+=====================+===========================+=============================+
+| :code:`first_open_paren`         | |values|  | opening parenthesis | |default_remove_new_line| | * |add_new_line|            |
++----------------------------------+-----------+---------------------+---------------------------+ * |remove_new_line|         |
+| :code:`last_close_paren`         | |values|  | closing parenthesis | |default_ignore|          | * |ignore|                  |
++----------------------------------+-----------+---------------------+---------------------------+                             |
+| :code:`interface_element`        | |values|  | interface element   | |default_add_new_line|    |                             |
```

Comment:
Thanks for catching this. If I remember correctly, the reason that this discrepancy came about is that I had initially set the default to `add_new_line`, but I found that it caused some unit tests for the JCL style to fail.

I would prefer to set it to `add_new_line`, since it is easier for users to spot a correction that they don't want and disable the rule than it is for them to look through the rules for one that is a disabled rule. As long as you're fine with me tinkering with the expected result for the JCL style unit tests, I'll change the default.

Reply:
> However, I worry about affecting the existing behaviour of the tool too much.

I think in this case, since the rules are new, we are free to set the default behavior.  I would propose:
```yaml
  function_019:
    first_open_paren: 'remove_new_line',
    last_close_paren: 'add_new_line',
    interface_list_semicolon: 'add_new_line',
    interface_element: 'add_new_line',
    ignore_single_line: 'no'
```
I think the only controversial options are `last_close_paren` and `ignore_signle_line`.

The `ignore_single_line` option is really an exception to the rule, so setting it to `no` would enforce consistent formatting.  Which I believe is a good thing.  This would force the user to change the option if they disagree.

I can see `last_close_paren` being either `add_new_line` or `remove_new_line` and I do not have a strong opinion on which it should be so I would leave it up to you to decide.

> However, if you are OK with the existing behaviour being affected, I will change them both to add_new_line and update the expected output for the JCL style unit tests.

If I add a new rule that starts failing the JCL style, I will change the style configuration so the tests pass (assuming that is possible).   Looking at the fixed files, the following configuration should work:
```yaml
  function_019:
    first_open_paren: 'remove_new_line',
    last_close_paren: 'add_new_line',
    interface_list_semicolon: 'add_new_line',
    interface_element: 'add_new_line',
    ignore_single_line: 'yes'
```

Honestly, I would like to remove the `jcl` style.  I never use it and it's only real purpose now is to help catch rules that interfere with each other.  I'm not convinced it does a good job at that either.  I just do not know if anybody else is using it.

The styles feature never turned out that way I hoped it would.  I was imagining people would submit their own styles as a pull request.  Then we could have a library of styles that someone could chose from similar to editor themes.  Unfortunately, nobody has submitted any.  When I reflect on it, it makes sense.  Submitting a PR for a style would mean it would need constant maintaining.  I try to not put in features nobody uses, but this must have been me thinking it is a cool feature but nobody else did.  So now I have to carry it around.  It's a little frustrating, but also a learning experience.

Reply:
Thanks very much for your detailed response. After giving it a try, I've followed your suggestions, except I used `remove_new_line` for `interface_list_semicolon` since it was failing the JCL style and it looked very strange to me to have the semicolon on its own line.

Yes, that is a shame! I can definitely see the appeal of having the library of styles, but I suppose people who don't care much about the style as long as it's consistent would tend just use the default settings, and people who are fussy would just set up their own config, and as you say, the maintainance work is not ideal. At the very least, it is handy for testing.

Reply:
Morning @JHertz5 ,

> except I used `remove_new_line` for `interface_list_semicolon`

I agree.  Not sure how I made that `add_new_line`.